## One sample t-Test

In [1]:
from scipy import stats

# Sample data (cookie weights)
cookie_weights = [29.8, 30.1, 30.2, 29.9, 30.0, 30.3]

# Perform one-sample t-test
t_stat, p_value = stats.ttest_1samp(cookie_weights, 30)
print(f"T-statistic: {t_stat}, P-value: {p_value}")


T-statistic: 0.6546536707079857, P-value: 0.5416045607931151


## Result

- A larger t-statistic means the sample mean is farther from the population mean.Here the t-statistic is small (~0.65), indicating that the sample mean is close to the population mean of 30 grams.
- A p-value greater than 0.05 means that we fail to reject the null hypothesis. Here the p-value is ~0.54, which is quite high. This suggests that the difference between the sample mean and the population mean of 30 grams is not statistically significant.

## Two-Sample t-Test

In [39]:
# Group A and Group B weight loss data
group_A = [2.1, 2.5, 2.8, 3.0, 3.2]
group_B = [1.8, 2.0, 2.1, 2.3, 2.5]

# Perform two-sample t-test
t_stat, p_value = stats.ttest_ind(group_A, group_B)
print(f"T-statistic: {t_stat}, P-value: {p_value}")


T-statistic: 2.5434682559903847, P-value: 0.03452387876553226


## Result
- The T-statistic of 2.54 indicates a significant difference in weight loss between Group A and Group B.
- The P-value of 0.0345 is less than the common significance level of 0.05. Therefore, we reject the null hypothesis, suggesting that there is statistically significant evidence that Group A experienced greater weight loss than Group B. This indicates that the method or treatment used in Group A is likely more effective in promoting weight loss compared to the method used in Group B.

## Paired t-Test

In [38]:
# Before and after test scores
before = [70, 75, 72, 78, 74]
after = [78, 80, 82, 85, 84]

# Perform paired t-test
t_stat, p_value = stats.ttest_rel(before, after)
print(f"T-statistic: {t_stat}, P-value: {p_value}")


T-statistic: -8.432740427115679, P-value: 0.0010829785357964236


## Result
- A large negative T-statistic -8.43 indicates a significant decrease in test scores before tutoring compared to after tutoring.
- The P-value 0.00108 is much smaller than 0.05, so we reject the null hypothesis.
- The extremely low P-value suggests that there is strong statistical evidence that the tutoring had a significant positive impact on the students' test scores. The negative T-statistic shows that the scores after tutoring are much higher than before tutoring. Therefore, the tutoring was effective in improving the students' performance.

## Chi-Square Test

In [40]:
import pandas as pd
from scipy.stats import chi2_contingency

# Contingency table: Gender vs Phone brand choice
data = pd.DataFrame({
    'Brand A': [30, 25],  # First row for Male, second row for Female
    'Brand B': [20, 35],
    'Brand C': [50, 40]
}, index=['Male', 'Female'])

# Perform Chi-Square test
chi2_stat, p_val, dof, ex = chi2_contingency(data)
print(f"Chi-Square Statistic: {chi2_stat}, P-value: {p_val}")


Chi-Square Statistic: 5.656565656565656, P-value: 0.05911427592751336


## Result

- Chi-Square Statistic 5.66 suggests there might be some association between gender and smartphone brand choice.
- The P-value 0.0591 is greater than the common significance level 0.05, so you fail to reject the null hypothesis. This means that, based on the data you have, the evidence isn't strong enough to claim a statistically significant association between gender and smartphone brand choice.

## ANOVA

In [42]:
# Data: Test scores of students in 3 groups
group_A = [85, 88, 90, 93, 87]
group_B = [78, 82, 85, 80, 81]
group_C = [92, 94, 89, 95, 93]

# Perform ANOVA
f_stat, p_value = stats.f_oneway(group_A, group_B, group_C)
print(f"F-statistic: {f_stat}, P-value: {p_value}")


F-statistic: 23.55868544600939, P-value: 6.995146378161676e-05


## Result
- The larger F-statistic 23.56 suggests that the differences between groups are large compared to the variability within groups.
- P-value: .0000699 is extremely small, much smaller than the typical significance level of 0.05.
- Since the P-value is much less than 0.05, we reject the null hypothesis. This means there is strong evidence to suggest that the teaching method significantly influences student test scores. At least one of the groups (A, B, or C) has a significantly different mean score compared to the others.

## Regression

In [3]:
from scipy.stats import linregress
x = [1, 2, 3, 4, 5]
y = [2, 4, 5, 4, 5]
slope, intercept, r_value, p_value, std_err = linregress(x, y)
print(f"Slope: {slope}, P-Value: {p_value}")

Slope: 0.6000000000000001, P-Value: 0.12402706265755442


## Result
While the slope indicates some positive relationship between x and y, the p-value suggests that this relationship is not statistically significant. Therefore, we cannot confidently claim a strong or reliable linear association based on this data alone.

## one sample z test

In [4]:
from statsmodels.stats.weightstats import ztest

# Sample data
sample_heights = [172, 171, 173, 170, 169, 172, 174, 171, 172, 170] * 5  # Total 50 data points

# Population parameters
population_mean = 170
population_std = 6  # known population std

# Perform Z-test
z_stat, p_value = ztest(sample_heights, value=population_mean, alternative='two-sided')

print(f"Z-statistic: {z_stat}, P-value: {p_value}")

Z-statistic: 6.861372411737276, P-value: 6.820210334434062e-12


## Result
-  Here the p-value is very small and Z-statistic is very high, so we can reject the null hypothesis. This indicates that the mean height of the sample is significantly different from the population mean of 170 cm.

## Two sample z test

In [5]:
from statsmodels.stats.weightstats import ztest

# Sample data
class_A_heights = [172, 171, 173, 170, 169, 172, 174, 171, 172, 170] * 10  # n=100
class_B_heights = [174, 175, 173, 176, 174, 173, 175, 176, 174, 172] * 12  # n=120

# Perform two-sample Z-test
z_stat, p_value = ztest(class_A_heights, class_B_heights, alternative='two-sided')

print(f"Z-statistic: {z_stat}, P-value: {p_value}")


Z-statistic: -15.437114936681608, P-value: 9.213013912296002e-54


## Result
- Here the p-value is very small and the Z-statistic is large, we reject the null hypothesis. This means there is strong evidence that the average heights of Class A and Class B are significantly different.

## Correlation

In [43]:
import numpy as np

# Data: Hours of study and corresponding test scores
hours = np.array([1, 2, 3, 4, 5])
scores = np.array([50, 55, 60, 70, 80])

# Compute correlation
correlation = np.corrcoef(hours, scores)
print(f"Correlation matrix:\n{correlation}")


Correlation matrix:
[[1.         0.98479825]
 [0.98479825 1.        ]]


## Result
- Correlation coefficient: 0.9848 This value is very close to 1, indicating a strong positive correlation between the number of hours studied and the test scores. As the hours of study increase, test scores tend to increase as well.

## Confidence Intervals

In [2]:
import numpy as np
import scipy.stats as stats
mean = 170
std_dev = 10
n = 100
confidence_interval = stats.norm.interval(0.95, loc=mean, scale=std_dev/np.sqrt(n))
print(confidence_interval)


(168.04003601545995, 171.95996398454005)


## Result
If you survey 100 people and find an average height of 170 cm, a 95% confidence interval might be 168-172 cm.

## Hypothesis testing

In [46]:
import numpy as np
from scipy import stats

# Test scores for both methods
traditional_scores = np.array([70, 75, 80, 68, 72, 78, 74, 71, 79, 73])
new_method_scores = np.array([85, 90, 88, 92, 86, 91, 89, 87, 93, 90])

# Perform the paired T-test
t_stat, p_value = stats.ttest_rel(traditional_scores, new_method_scores)

print(f"T-statistic: {t_stat}, P-value: {p_value}")


T-statistic: -12.068205739876403, P-value: 7.336865193961798e-07


## Result 

- The T-statistic −12.07 indicates that there is a significant difference between the scores from the traditional method and the new method, with the new method generally yielding higher scores.
- The P-value 0.0000007337 is extremely small, much less than the common significance level of 0.05.
-  Since the P-value is much less than 0.05, we reject the null hypothesis. This indicates that there is strong evidence that the new teaching method has a significantly positive impact on student test scores compared to the traditional method. In other words, students scored significantly higher with the new method.

In [21]:
## Example

In [47]:
import pandas as pd
import numpy as np

# Set random seed for reproducibility
np.random.seed(42)

# Number of students
n_students = 200

# Generate data
student_ids = np.arange(1, n_students + 1)
ages = np.random.randint(18, 25, size=n_students)
genders = np.random.choice(['Male', 'Female'], size=n_students, p=[0.5, 0.5])
study_hours = np.random.normal(10, 3, size=n_students).round(1)  # mean=10, std=3
study_hours = np.clip(study_hours, 1, 20)  # Set limits on study hours
test_scores = np.random.normal(70, 15, size=n_students).round(1)  # mean=70, std=15
test_scores = np.clip(test_scores, 40, 100)  # Limit test scores between 40 and 100
tutoring = np.random.choice([0, 1], size=n_students, p=[0.7, 0.3])  # 30% students get tutoring
courses = np.random.choice(['Science', 'Arts', 'Commerce'], size=n_students, p=[0.4, 0.3, 0.3])
extracurricular = np.random.choice([0, 1], size=n_students, p=[0.6, 0.4])  # 40% participate in extracurricular activities
diet_type = np.random.choice(['Vegetarian', 'Non-Vegetarian', 'Vegan'], size=n_students, p=[0.3, 0.5, 0.2])
satisfaction = np.random.randint(1, 6, size=n_students)  # Satisfaction rating from 1 to 5

# Create DataFrame
data = pd.DataFrame({
    'Student_ID': student_ids,
    'Age': ages,
    'Gender': genders,
    'Study_Hours_per_Week': study_hours,
    'Test_Score': test_scores,
    'Tutoring': tutoring,
    'Course': courses,
    'Extracurricular': extracurricular,
    'Diet_Type': diet_type,
    'Satisfaction': satisfaction
})

# Show first 5 rows of the dataset



In [49]:
data.head(10)

,Student_ID,Age,Gender,Study_Hours_per_Week,Test_Score,Tutoring,Course,Extracurricular,Diet_Type,Satisfaction
0,1,24,Male,9.6,40.0,1,Commerce,1,Non-Vegetarian,1
1,2,21,Male,7.6,99.4,0,Science,1,Vegetarian,2
2,3,22,Male,8.9,66.1,0,Arts,0,Non-Vegetarian,1
3,4,24,Male,10.5,78.0,1,Arts,0,Vegan,4
4,5,20,Male,9.7,98.6,0,Arts,0,Non-Vegetarian,1
5,6,22,Female,12.5,73.4,0,Arts,1,Vegetarian,4
6,7,22,Male,9.0,67.1,1,Science,1,Non-Vegetarian,4
7,8,24,Male,9.4,91.3,0,Arts,0,Non-Vegetarian,3
8,9,19,Male,10.8,76.9,0,Arts,1,Non-Vegetarian,5
9,10,20,Female,16.7,78.8,0,Commerce,1,Vegetarian,4


### One-Sample t-Test:
- Hypothesis: Test whether the average test score is equal to 75.

In [58]:
from scipy import stats
t_stat, p_value = stats.ttest_1samp(data['Test_Score'], 75)
print(f"One-Sample t-Test P-value: {p_value}")

One-Sample t-Test P-value: 0.004688259109513488


### Two-Sample t-Test:
- Hypothesis: Compare test scores of students with and without tutoring.

In [59]:
tutored = data[data['Tutoring'] == 1]['Test_Score']
not_tutored = data[data['Tutoring'] == 0]['Test_Score']
t_stat, p_value = stats.ttest_ind(tutored, not_tutored)
print(f"Two-Sample t-Test P-value: {p_value}")


Two-Sample t-Test P-value: 0.3569884881420057


### ANOVA:
- Hypothesis: Compare mean test scores across different courses.

In [60]:
f_stat, p_value = stats.f_oneway(data[data['Course'] == 'Science']['Test_Score'],
                                 data[data['Course'] == 'Arts']['Test_Score'],
                                 data[data['Course'] == 'Commerce']['Test_Score'])
print(f"ANOVA P-value: {p_value}")

ANOVA P-value: 0.6826094780736212


### Chi-Square Test:
- Hypothesis: Check if gender is associated with the course selected.

In [61]:
contingency_table = pd.crosstab(data['Gender'], data['Course'])
chi2_stat, p_value, dof, ex = stats.chi2_contingency(contingency_table)
print(f"Chi-Square P-value: {p_value}")


Chi-Square P-value: 0.7930725783353905


### Correlation:
- Hypothesis: Check the relationship between study hours and test scores.

In [62]:
correlation, _ = stats.pearsonr(data['Study_Hours_per_Week'], data['Test_Score'])
print(f"Correlation between Study Hours and Test Score: {correlation}")


Correlation between Study Hours and Test Score: 0.10862589385920665


##  Result:

- One-Sample t-Test
    - Hypothesis: Test whether the average test score is equal to 75.
    - The P-value .0047 is less than 0.05, we reject the null hypothesis, indicating that the average test score is significantly different from 75.

- Two-Sample t-Test
    - Hypothesis: Compare test scores of students with and without tutoring.
    - The P-value .3570 is greater than 0.05, suggesting no significant difference in test scores between the two groups (tutored vs. not tutored). We fail to reject the null hypothesis, indicating that tutoring does not have a significant effect on test scores.

- ANOVA
    - Hypothesis: Compare mean test scores across different courses.
    - The P-value .6826 is much greater than 0.05, indicating no significant differences in mean test scores across the different courses. We fail to reject the null hypothesis, suggesting that the teaching methods (or courses) do not significantly affect test scores.


- Chi-Square Test
    - Hypothesis: Check if gender is associated with the course selected.
    - The high P-value .7931 indicates no significant association between gender and the selected course. We fail to reject the null hypothesis, suggesting that gender does not influence course selection.


- Correlation
    - Hypothesis: Check the relationship between study hours and test scores.
    - The low positive correlation 0.109 suggests a very weak relationship between study hours and test scores, indicating that increasing study hours does not significantly impact academic performance.

#### Overall Summary
- The One-Sample t-Test indicates that the average test score differs from 75.
- The Two-Sample t-Test and ANOVA show that tutoring and course selection do not significantly impact test scores.
- The Chi-Square Test reveals no association between gender and course selection.
- The Correlation analysis suggests that study hours have a weak predictive relationship with test scores.